Update the Android build system

mmatyas opened the pull request:
A huge update for the Android build system, which makes the Android target build again.

There are still some runtime issues, see #13154. CC & created by @larsbergstrom.

Fixes #12562.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13945)

<!-- Reviewable:end -->

mmatyas committed:
Update to use the new Android library build rules.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/build_commands.py

Ms2ger commented:
r? @larsbergstrom

aneeshusa commented on components/servo/build.rs:
```diff
@@ -17,22 +17,46 @@ fn main() {
-    let standalone_path = env::var("NDK_STANDALONE").ok().expect("Please set the NDK_STANDALONE environment variable");
-    let standalone_path = Path::new(&standalone_path);
+    // Build up the path to the NDK compilers
+    // Options for host are:  "linux-x86_64" "linux-x86" "darwin-x86_64" "darwin-x86"
```
Is `darwin-x86` a relevant option? I don't see it listed below, although it is mentioned in other places.

larsbergstrom replied:
I guess we don't actually build for Mac 32-bit :-)

aneeshusa commented on components/servo/fake-ld.sh:
```diff
@@ -9,7 +9,25 @@ set -o nounset
-touch "${TARGET_DIR}/servo"
+
+export _ANDROID_ARCH=arch-arm
+export ANDROID_SYSROOT="$ANDROID_NDK/platforms/android-18/$_ANDROID_ARCH"
```
Please use the full `"${ANDROID_NDK}/platforms ...` quoting style per the [style guide](https://github.com/servo/servo/blob/master/docs/STYLE_GUIDE.md)

aneeshusa commented on components/servo/fake-ld.sh:
```diff
@@ -9,7 +9,25 @@ set -o nounset
+export ANDROID_SYSROOT="$ANDROID_NDK/platforms/android-18/$_ANDROID_ARCH"
+export _ANDROID_EABI=arm-linux-androideabi-4.9
+ANDROID_TOOLCHAIN=""
+for host in "linux-x86_64" "linux-x86" "darwin-x86_64" "darwin-x86"
```
style nit: use a `; do` at the end of this line instead of putting it on the next line

aneeshusa commented on python/servo/build_commands.py:
```diff
@@ -244,6 +246,38 @@ def build(self, target=None, release=False, dev=False, jobs=None,
+            os_type = platform.system().lower()
+            cpu_type = platform.machine().lower()
+            if os_type == "linux":
+                if cpu_type in ["i386", "i486", "i686", "i768", "x86"]:
```
You only need to have this block once to find the host suffix (e.g. `x86_64`), and then can append it to `os_type + '-'` directly.

aneeshusa commented on components/servo/fake-ld.sh:
```diff
@@ -9,7 +9,25 @@ set -o nounset
+ANDROID_CXX_LIBS=$ANDROID_NDK/sources/cxx-stl/llvm-libc++/libs/$ANDROID_CPU_ARCH_DIR
+echo libs dir: $ANDROID_CXX_LIBS
+
+echo sysroot: $ANDROID_SYSROOT
```
style nit: The `echo` calls looks like they're left over from debugging but may still be useful; if you keep them please quote the whole argument to `echo`.

mmatyas commented:
Ok, here's an update. I've added quotes in `fake-ld.sh`, though `mach test-tidy` still throws warnings; I wonder if that's a bug or did I miss something? I've also changed the indentation of the paths and `C/CXXFLAGS' in `build_commands.py`, tell me what do you think.

larsbergstrom commented:
```
./components/servo/fake-ld.sh:18: Line is longer than 80 characters

./components/servo/fake-ld.sh:19: Line is longer than 80 characters

./components/servo/fake-ld.sh:25: Line is longer than 80 characters

./components/servo/fake-ld.sh:32: Line is longer than 80 characters

./components/servo/fake-ld.sh:18: variable substitutions should use the full "${VAR}" form

./components/servo/fake-ld.sh:19: variable substitutions should use the full "${VAR}" form

./components/servo/fake-ld.sh:32: variable substitutions should use the full "${VAR}" form

./python/servo/build_commands.py:251: Line is longer than 120 characters
```

Other than the errors from tidy, this looks good. The line length ones seem legit, but I'm not sure what's going on with the substitution errors...

mmatyas commented:
I just noticed I've forgot to put `$host` between `{}`.

aneeshusa commented on components/servo/fake-ld.sh:
```diff
@@ -9,7 +9,25 @@ set -o nounset
+echo "sysroot: ${ANDROID_SYSROOT}"
+
+"${ANDROID_TOOLCHAIN}/arm-linux-androideabi-gcc" \
+  --sysroot="${ANDROID_SYSROOT}" -L "${ANDROID_CXX_LIBS}" $@ ${LDFLAGS:-} -lc++ \
```
I think tidy is looking for `"${@}"` to be quoted.

Also, can you see if `$LDFLAGS` is required? See #12562.

mmatyas replied:
Ah my bad, I've forget the `{}`s, tidy does acccept it now.

aneeshusa replied:
I'm pretty sure this script is only for Android, so in that case I'm comfortable removing `LDFLAGS`.

mmatyas commented:
As for the last `"${VAR}"` warning, I think that's for the `LDFLAGS`, but that's probably fine.

mmatyas commented:
I've double checked just in case, but yes, `LDFLAGS` is unbound, so I've removed it now. I've also merged my fixups into one commit.

mmatyas commented:
I've fixed the tidy errors: I've removed the unnecessary bits from the URL in `build_commands.py`, and added `fake-ld.sh` to the ignore list, as discussed on IRC.

larsbergstrom commented:
@bors-servo r=larsbergstrom,aneeshusa

bors-servo commented:
:pushpin: Commit 5bb550d has been approved by `larsbergstrom,aneeshusa`

<!-- @bors-servo r=larsbergstrom,aneeshusa 5bb550dbf03946578344897753dbec5d836b348a -->

bors-servo commented:
:hourglass: Testing commit 5bb550d with merge b933015...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3769), [arm64](http://build.servo.org/builders/arm64/builds/3779), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3853), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/494), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/492), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3847), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3837), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/528), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/532), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1757)
